blog/en/blog/2023/07/28/request-validation.md

```diff
@@ -0,0 +1,410 @@
+  "files": {},
+  "form": {},
+  "headers": {
+    "Accept": "*/*",
+    "Content-Length": "53",
+    "Content-Type": "application/json",
+    "Host": "localhost:9080",
+    "User-Agent": "curl/7.88.1",
+    "X-Forwarded-Host": "localhost"
+  },
+  "json": {
+    "item": "sticker",
+    "price": 1,
+    "quantity": 20,
+    "uid": 4639
+  },
+  "method": "POST",
+  "origin": "172.22.0.1, 223.239.0.41",
+  "url": "http://localhost/anything/anything"
+}
+```
+
+This [excellent blog post](https://bishopfox.com/blog/json-interoperability-vulnerabilities) sums up some more JSON interoperability vulnerabilities.
+
+## Complex Validation
+
+What we discussed so far only included some basic validation. But APISIX has much more capabilities.
+
+You can also configure a schema for validating headers, set a custom response code/message, and validate strings using regular expressions:
+
+```yaml title="apisix.yaml" {9-10,11,18}
+routes:
+  - uri: /anything/*
+    upstream:
+      type: roundrobin
+      nodes:
+        "httpbin:80": 1
+    plugins:
+      request-validation:
+        rejected_code: 418
+        rejected_message: "I'm an intelligent Teapot!"
+        header_schema:
+            type: object
+            required:
+              - Content-Type
+            properties:
+              Content-Type:
+                type: string
+                pattern: "^application\/json$"
+        body_schema:
+            type: object
+            required:
+              - quantity
+              - price
+              - item
+            properties:
+              uid:
+                type: integer
+              item:
+                type: string
+                enum:
+                  - sticker
+                  - t-shirt
+                  - pin
+              quantity:
+                type: integer
+                minimum: 1
+                maximum: 50
+              price:
+                type: number
+                minimum: 1.0
+                maximum: 50.0
+#END
+```
+
+You can also set the specific version of the JSON schema used from draft 4, draft 6, and draft 7 to ensure accurate parsing by adding this key:
+
+```text
+"$schema": "http://json-schema.org/draft-04/schema#"
+```
+
+## ~~Why~~ Where to Validate?
+
+It is fair to assume that you now understand why request validation might be necessary. However, you could still argue that validation can be done on the client side or within your backend services.
+
+A key point is that you might not always own your clients, or modifying client requests might be relatively easy. Adding validation in a layer you have absolute control of is the better alternative.
+
+So why can't you validate requests in your services directly? Well, after a point, it becomes too complex for the application developer to configure both request validation and application/business-specific validation. Without an API gateway layer, invalid requests still end up adding load to your backend services.
```

Comment:
On that, I completely disagree. On the opposite, I think that's the main argument you need to make in the introduction.

Validation is business logic. You want to avoid as much business logic as possible in the gateway because of deployment synchronization issue. Putting something as trivial as body validation in the API Gateway causes sync problems for not much value - a trip from the gateway to the upstream (they should be close anyway).

Validating headers, however, is much more useful (for security) and is much less business related.

Reply:
It is mentioned in the introduction. A trip from the gateway to the upstream is not the issue, but sending invalid requests to the upstream unnecessarily can be avoided.